sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -274,7 +275,21 @@ object InMemoryFileIndex extends Logging {
     // [SPARK-17599] Prevent InMemoryFileIndex from failing if path doesn't exist
     // Note that statuses only include FileStatus for the files and dirs directly under path,
     // and does not include anything else recursively.
-    val statuses = try fs.listStatus(path) catch {
+    val statuses: Array[FileStatus] = try {
+      fs match {
+        // DistributedFileSystem overrides listLocatedStatus to make 1 single call to namenode
+        // to retrieve the file status with the file block location. The reason to still fallback
+        // to listStatus is because the default implementation would potentially throw a
+        // FileNotFoundException which is better handled by doing the lookups manually below.
```

Comment:
FYI, over at https://github.com/apache/spark/pull/24668 I'm proposing a change to the `FileNotFoundException` handling logic here: after my patch's changes, we'll only ignore `FileNotFoundException` if (a) we are not listing the root directory of the table (i.e. we are listing a subfolder) and (b) the `ignoreFile` setting has not been set.

Does that change have any implications for this PR?

Reply:
I believe that should not matter here as I tried to preserve the original logic (which would be your new logic now) when using listStatus which has that issue. The only real change here is using the DistributedFileSystem.listLocatedStatus and there should be no case where FileNotFoundException happens when using that as they will all be LocatedFileStatus that are returned in the array.

Reply:
Ah, I think I get it now: I was originally confused by the comment, but on re-read it sounds like you're saying that we _could_, in principle, just call `listLocatedStatus()` in all cases, except that there's no performance benefit for non `DistributedFileSystem` subclasses and we lose more granular error handling.

In fact, the granular error handling could be really important in order to preserve the behavior before my changes in #24668 (a.k.a. the new behavior when `ignoreMissingFiles = true`): previously, the deletion of a leaf file after a directory listing but before listing the file status would cause _only that file_ to be ignored, whereas using `listLocatedStatus` for non-`DistributedFileSystem` subclasses would allow that `FileNotFoundException` to trigger the dropping of _all_ files from the listing (even ones whose statuses could be retrieved).

I feel like `DistributedFileSystem` must be having some sort of internal "atomicity"-type guarantee for this to be safe, though (that it if it sees a file then it's able to also see its status).

Reply:
Yeah these were my thoughts exactly. My first version of this removed error handling logic and just used listLocatedStatus instead but when i inspected the source I realized we'd definitely run into issues with discarding the entire directory with 1 file not found.

Reply:
👍 

Based on this discussion I've strengthened my test assertions to guard against this problem: https://github.com/apache/spark/pull/24668/commits/42e8b9881d2654150a92bbc7fc341904f4862900. The tests will now fail in case `listLocatedStatus()` is always unconditionally used.